File: iznik-server-go/ormharness/resultparity.go
package ormharness

// Layer 2 of the ORM migration harness (plan section 7.2): result parity on
// fixtures. Layer 1 (canonical.go, golden.go - owned elsewhere in this
// package) proves the new GORM call renders equivalent SQL text. This file
// proves it independently, by actually running both the old raw SQL and the
// new GORM call against the seeded test database and diffing the result
// sets. That catches what string comparison cannot: collation, NULL
// handling and implicit-cast semantics that only show up once the query
// hits the engine.

import (
	"database/sql"
	"fmt"
	"reflect"
	"regexp"
	"sort"
	"strings"
	"testing"

	"gorm.io/gorm"
)

// ReplacementQuery builds, but does not execute, the GORM replacement for a
// raw SQL call site under test - e.g.
//
//	func(tx *gorm.DB) *gorm.DB {
//		return tx.Table("users").Select("id, email").Where("id = ?", 5)
//	}
//
// It must stop short of a terminal method (Find, First, Scan, Row, Rows...):
// AssertResultParity calls Rows() itself so that it can scan both result
// sets through the same generic path.
type ReplacementQuery func(tx *gorm.DB) *gorm.DB

// resultRow is one scanned row, keyed by column name. Values come straight
// out of database/sql with no destination type imposed, so a NULL is a true
// nil, and the concrete Go type reflects whatever the driver decided the
// column's SQL type maps to (int64, float64, []byte, time.Time...). That is
// what makes the comparison NULL-sensitive and cast-sensitive: a NULL can
// never equal "" or 0, and a column that comes back as []byte on one side
// but int64 on the other (an implicit-cast divergence) fails reflect.DeepEqual
// even when its printed form looks the same.
type resultRow map[string]any

// orderByPattern flags a statement as having its own ordering. It is
// deliberately simple (the plan's wording is "contains ORDER BY"): a
// case-insensitive word-boundary match is enough for real call-site SQL,
// and false positives inside a string literal would only make the harness
// stricter (order-sensitive) rather than silently lenient.
var orderByPattern = regexp.MustCompile(`(?is)\border\s+by\b`)

// AssertResultParity runs originalSQL (with args) and replacement against
// db and fails t if the two result sets are not identical: same columns
// (as a set), same row count, and every value equal including its NULL-ness
// and its Go type. When originalSQL contains ORDER BY the rows are compared
// position by position, since the two implementations are then contractually
// obliged to agree on order; otherwise both result sets are sorted into a
// shared canonical order before comparing, because without ORDER BY the
// engine is free to return rows in any order and the old and new query need
// not pick the same one.
func AssertResultParity(t *testing.T, db *gorm.DB, originalSQL string, args []any, replacement ReplacementQuery) {
	t.Helper()

	originalCols, originalRows, err := runRawSQL(db, originalSQL, args)
	if err != nil {
		t.Fatalf("ormharness: original SQL failed: %v\nSQL: %s", err, originalSQL)
	}

	replacementCols, replacementRows, err := runReplacementQuery(db, replacement)
	if err != nil {
		t.Fatalf("ormharness: replacement GORM call failed: %v", err)
	}

	assertSameColumnSet(t, originalCols, replacementCols)

	if orderByPattern.MatchString(originalSQL) {
		assertRowsEqualOrdered(t, originalCols, originalRows, replacementRows)
	} else {
		assertRowsEqualUnordered(t, originalCols, originalRows, replacementRows)
	}
}

// runRawSQL executes the golden raw SQL through GORM's Raw/Rows path (the
// same driver, connection pool and dbresolver routing as the ORM side gets),
// so any difference found downstream is attributable to the query, not to
// how it was run.
func runRawSQL(db *gorm.DB, sqlText string, args []any) ([]string, []resultRow, error) {
	rows, err := db.Raw(sqlText, args...).Rows()
	if err != nil {
		return nil, nil, err
	}
	defer rows.Close()

	return scanRowsGenerically(rows)
}

// runReplacementQuery executes the built-but-unexecuted GORM chain the
// caller supplied.
func runReplacementQuery(db *gorm.DB, replacement ReplacementQuery) ([]string, []resultRow, error) {
	rows, err := replacement(db).Rows()
	if err != nil {
		return nil, nil, err
	}
	defer rows.Close()

	return scanRowsGenerically(rows)
}

// scanRowsGenerically reads every row of rows into a resultRow keyed by
// column name, without knowing the shape of the query in advance. Scanning
// into *any destinations (rather than a typed struct or sql.RawBytes) makes
// database/sql's MySQL driver hand back its native representation per
// column - nil for NULL, int64/float64/[]byte/time.Time otherwise - which is
// exactly the type-preserving, NULL-preserving behaviour Layer 2 needs and
// works for any site's result shape with no per-site scaffolding.
func scanRowsGenerically(rows *sql.Rows) ([]string, []resultRow, error) {
	cols, err := rows.Columns()
	if err != nil {
		return nil, nil, err
	}

	var out []resultRow
	for rows.Next() {
		raw := make([]any, len(cols))
		ptrs := make([]any, len(cols))
		for i := range raw {
			ptrs[i] = &raw[i]
		}
		if err := rows.Scan(ptrs...); err != nil {
			return nil, nil, err
		}

		row := make(resultRow, len(cols))
		for i, c := range cols {
			row[c] = raw[i]
		}
		out = append(out, row)
	}
	if err := rows.Err(); err != nil {
		return nil, nil, err
	}

	return cols, out, nil
}

// assertSameColumnSet fails t if the two queries did not select the same
// columns. Compared as a set (sorted) rather than positionally: Layer 1
// already pins the SQL text (and therefore column order) for converted
// sites, so Layer 2's job is to catch data divergence, not re-litigate
// column ordering.
func assertSameColumnSet(t *testing.T, originalCols, replacementCols []string) {
	t.Helper()

	a := append([]string(nil), originalCols...)
	b := append([]string(nil), replacementCols...)
	sort.Strings(a)
	sort.Strings(b)

	if !reflect.DeepEqual(a, b) {
		t.Fatalf("ormharness: column sets differ\n  original:    %v\n  replacement: %v", originalCols, replacementCols)
	}
}

// assertRowsEqualOrdered compares two result sets position by position.
func assertRowsEqualOrdered(t *testing.T, cols []string, original, replacement []resultRow) {
	t.Helper()

	if len(original) != len(replacement) {
		t.Fatalf("ormharness: row count differs (order-sensitive comparison): original=%d replacement=%d", len(original), len(replacement))
	}

	for i := range original {
		assertRowEqual(t, cols, fmt.Sprintf("row %d", i), original[i], replacement[i])
	}
}

// assertRowsEqualUnordered compares two result sets as multisets: both are
// sorted into the same canonical order (a deterministic string encoding of
// each row, so equal rows always sort adjacently and identically on both
// sides) before comparing position by position. This is only a sort key,
// not the equality test itself - assertRowEqual still does the real,
// type-and-NULL-sensitive comparison once the rows are lined up.
func assertRowsEqualUnordered(t *testing.T, cols []string, original, replacement []resultRow) {
	t.Helper()

	if len(original) != len(replacement) {
		t.Fatalf("ormharness: row count differs (unordered comparison): original=%d replacement=%d", len(original), len(replacement))
	}

	sortedOriginal := sortRowsCanonically(cols, original)
	sortedReplacement := sortRowsCanonically(cols, replacement)

	for i := range sortedOriginal {
		assertRowEqual(t, cols, fmt.Sprintf("row %d (canonical order, no ORDER BY on the original query)", i), sortedOriginal[i], sortedReplacement[i])
	}
}

func sortRowsCanonically(cols []string, rows []resultRow) []resultRow {
	out := append([]resultRow(nil), rows...)
	sort.SliceStable(out, func(i, j int) bool {
		return canonicalRowKey(cols, out[i]) < canonicalRowKey(cols, out[j])
	})
	return out
}

// canonicalRowKey renders a row as a single string for sorting purposes
// only - it does not need to be collision-free between genuinely different
// rows in general, only stable and applied identically to both result sets,
// which is enough to line up two multisets that are actually permutations
// of each other. formatValue's NULL/type distinctions carry through, so
// e.g. a NULL score and a zero score still sort (and later compare)
// differently.
func canonicalRowKey(cols []string, row resultRow) string {
	var b strings.Builder
	for _, c := range cols {
		b.WriteString(c)
		b.WriteByte('=')
		b.WriteString(formatValue(row[c]))
		b.WriteByte('\x1f') // unit separator: cheap, unlikely-to-collide field delimiter
	}
	return b.String()
}

// assertRowEqual compares a single row's values column by column, using
// reflect.DeepEqual so that a NULL (nil) never matches "" or 0, and a value
// that came back as a different Go type on each side (an implicit-cast
// divergence) is caught even when the two would print the same.
func assertRowEqual(t *testing.T, cols []string, rowLabel string, original, replacement resultRow) {
	t.Helper()

	for _, c := range cols {
		originalValue, replacementValue := original[c], replacement[c]
		if !reflect.DeepEqual(originalValue, replacementValue) {
			t.Fatalf("ormharness: %s, column %q differs:\n  original:    %s\n  replacement: %s",
				rowLabel, c, formatValue(originalValue), formatValue(replacementValue))
		}
	}
}

// formatValue renders a scanned column value for failure messages (and as
// the building block for canonicalRowKey), keeping NULL visually distinct
// from an empty string or a zero value.
func formatValue(v any) string {
	if v == nil {
		return "NULL"
	}

	if b, ok := v.([]byte); ok {
		return fmt.Sprintf("%q (%T)", string(b), v)
	}

	return fmt.Sprintf("%v (%T)", v, v)
}

File: iznik-batch/tests/Support/OrmHarness/ResultParity.php
<?php

namespace Tests\Support\OrmHarness;

use Illuminate\Database\Eloquent\Builder as EloquentBuilder;
use Illuminate\Database\Query\Builder as QueryBuilder;
use Illuminate\Support\Facades\DB;
use PDO;
use PHPUnit\Framework\Assert;

/**
 * Layer 2 of the Laravel ORM migration harness (plan section 7.2): runs the
 * original raw SQL and the replacement query-builder call against the
 * seeded test database and diffs the result sets. Catches what Layer 1's
 * string comparison cannot - collation, NULL handling, implicit-cast
 * semantics that only show up once a query hits the engine. Design ported
 * from iznik-server-go/ormharness/resultparity.go's AssertResultParity;
 * read that file's own doc comment for the parts that transfer unchanged
 * (order-sensitive comparison only when the original SQL contains ORDER BY,
 * unordered/canonical-sorted comparison otherwise, because without ORDER BY
 * the engine is free to return rows in any order and old/new need not agree
 * on one).
 *
 * ONE RUNNER FOR BOTH SIDES, BY DESIGN - THE ONE GENUINE SIMPLIFICATION
 * VERSUS resultparity.go. Go's version runs the original through
 * db.Raw(...).Rows() and the replacement through the GORM chain's OWN
 * .Rows(), two different code paths that happen to share one *sql.DB
 * connection pool - and because of that, roughly a third of resultparity.go
 * (protocolHint, renderForComparison) exists purely to explain a real
 * go-sql-driver/mysql quirk: the driver returns []byte for every column
 * under MySQL's TEXT protocol but native Go types under the BINARY
 * protocol, and which protocol a statement gets depends on whether it was
 * prepared, which in turn depends on whether it has arguments - so a
 * statement with bind args and the "same" statement without can come back
 * with different Go types for equal data. This class does not need that
 * defence, and not by accident: the replacement is rendered via
 * toSql()/getBindings() and then run through runRawSql() - THE SAME
 * function, same PDO instance, same fetch settings - that runs the
 * original. There is only one code path, so there is no second protocol to
 * diverge from the first; a type or NULL difference here can only mean the
 * two statements are genuinely different, never an artefact of how each was
 * executed.
 *
 * WHY QUERY\BUILDER, NOT ELOQUENT MODEL HYDRATION. Go's Layer 2 scans BOTH
 * sides generically via database/sql's *sql.Rows, deliberately not through
 * a GORM-scanned struct - see scanRowsGenerically's own comment: it wants
 * the driver's native, unmodified representation on both sides. The same
 * reasoning rules out fetching the replacement through Eloquent's
 * $model->attributesToArray() (which applies $casts) here: a model
 * attribute cast (e.g. 'id' => 'integer') could turn a string PDO returns
 * into a genuine int, an artificial divergence from the RAW query's output
 * that has nothing to do with whether the two statements select the same
 * data. assert()/assertForSite() normalise any Eloquent\Builder to its
 * underlying Query\Builder (->toBase()) before rendering, for the same
 * reason GoldenSql.php does.
 */
final class ResultParity
{
    /**
     * @param  callable(): QueryBuilder|EloquentBuilder  $replacementBuild
     */
    public static function assert(string $originalSql, array $bindings, callable $replacementBuild): void
    {
        [$originalCols, $originalRows] = self::runRawSql($originalSql, $bindings);

        $built = $replacementBuild();
        $query = $built instanceof EloquentBuilder ? $built->toBase() : $built;
        if (! $query instanceof QueryBuilder) {
            Assert::fail(sprintf(
                'ormharness: replacementBuild() must return a Query\Builder or Eloquent\Builder, got %s',
                get_debug_type($built)
            ));

            return;
        }
        [$replacementCols, $replacementRows] = self::runRawSql($query->toSql(), $query->getBindings());

        self::assertSameColumnSet($originalCols, $replacementCols);

        if (preg_match('/\border\s+by\b/i', $originalSql) === 1) {
            self::assertRowsEqualOrdered($originalCols, $originalRows, $replacementRows);
        } else {
            self::assertRowsEqualUnordered($originalCols, $originalRows, $replacementRows);
        }
    }

    /**
     * AssertResultParityForSite equivalent: same as assert(), but verifies
     * $siteId is a real manifest entry first, so a typo'd site id fails
     * loudly rather than quietly proving nothing. Exists for the same
     * reason golden.go's AssertResultParityForSite does - a site whose only
     * proof is a Layer 2 test has no other place its id gets checked
     * mechanically against the manifest.
     *
     * @param  callable(): QueryBuilder|EloquentBuilder  $replacementBuild
     */
    public static function assertForSite(string $siteId, string $originalSql, array $bindings, callable $replacementBuild): void
    {
        Manifest::site($siteId); // throws if $siteId is not real; see Manifest::site()
        self::assert($originalSql, $bindings, $replacementBuild);
    }

    /**
     * Runs $sql through the SAME low-level path regardless of whether it is
     * the hand-written original or the query-builder's own rendered SQL -
     * see this class's header for why that single-path design is what
     * removes the protocol-divergence class of false positive Go's
     * equivalent has to explain away.
     *
     * @return array{0:list<string>,1:list<array<string,mixed>>}
     */
    private static function runRawSql(string $sql, array $bindings): array
    {
        $pdo = DB::connection()->getPdo();
        $stmt = $pdo->prepare($sql);
        $stmt->execute(array_values($bindings));

        $columns = [];
        for ($i = 0; $i < $stmt->columnCount(); $i++) {
            $meta = $stmt->getColumnMeta($i);
            $columns[] = $meta['name'] ?? "col{$i}";
        }

        /** @var list<array<string,mixed>> $rows */
        $rows = $stmt->fetchAll(PDO::FETCH_ASSOC);
        $stmt->closeCursor();

        return [$columns, $rows];
    }

    /**
     * @param  list<string>  $originalCols
     * @param  list<string>  $replacementCols
     */
    private static function assertSameColumnSet(array $originalCols, array $replacementCols): void
    {
        $a = $originalCols;
        $b = $replacementCols;
        sort($a);
        sort($b);

        Assert::assertSame($a, $b, sprintf(
            "ormharness: column sets differ\n  original:    %s\n  replacement: %s",
            implode(', ', $originalCols),
            implode(', ', $replacementCols)
        ));
    }

    /**
     * @param  list<string>  $cols
     * @param  list<array<string,mixed>>  $original
     * @param  list<array<string,mixed>>  $replacement
     */
    private static function assertRowsEqualOrdered(array $cols, array $original, array $replacement): void
    {
        Assert::assertCount(
            count($original),
            $replacement,
            sprintf('ormharness: row count differs (order-sensitive comparison): original=%d replacement=%d', count($original), count($replacement))
        );

        foreach ($original as $i => $row) {
            self::assertRowEqual($cols, "row {$i}", $row, $replacement[$i]);
        }
    }

    /**
     * Both result sets are sorted into the same canonical order (a
     * deterministic string encoding of each row, so equal rows sort
     * adjacently and identically on both sides) before comparing position
     * by position. This is only a sort key, not the equality test itself -
     * assertRowEqual still does the real, type-and-NULL-sensitive
     * comparison once rows are lined up.
     *
     * @param  list<string>  $cols
     * @param  list<array<string,mixed>>  $original
     * @param  list<array<string,mixed>>  $replacement
     */
    private static function assertRowsEqualUnordered(array $cols, array $original, array $replacement): void
    {
        Assert::assertCount(
            count($original),
            $replacement,
            sprintf('ormharness: row count differs (unordered comparison): original=%d replacement=%d', count($original), count($replacement))
        );

        $sortedOriginal = self::sortRowsCanonically($cols, $original);
        $sortedReplacement = self::sortRowsCanonically($cols, $replacement);

        foreach ($sortedOriginal as $i => $row) {
            self::assertRowEqual($cols, "row {$i} (canonical order, no ORDER BY on the original query)", $row, $sortedReplacement[$i]);
        }
    }

    /**
     * @param  list<string>  $cols
     * @param  list<array<string,mixed>>  $rows
     * @return list<array<string,mixed>>
     */
    private static function sortRowsCanonically(array $cols, array $rows): array
    {
        $out = $rows;
        usort($out, fn ($a, $b) => self::canonicalRowKey($cols, $a) <=> self::canonicalRowKey($cols, $b));

        return $out;
    }

    /**
     * @param  list<string>  $cols
     * @param  array<string,mixed>  $row
     */
    private static function canonicalRowKey(array $cols, array $row): string
    {
        $parts = [];
        foreach ($cols as $c) {
            $parts[] = $c.'='.self::formatValue($row[$c] ?? null);
        }

        return implode("\x1f", $parts); // unit separator: cheap, unlikely-to-collide delimiter
    }

    /**
     * @param  list<string>  $cols
     * @param  array<string,mixed>  $original
     * @param  array<string,mixed>  $replacement
     */
    private static function assertRowEqual(array $cols, string $rowLabel, array $original, array $replacement): void
    {
        foreach ($cols as $c) {
            $originalValue = $original[$c] ?? null;
            $replacementValue = $replacement[$c] ?? null;

            // Strict (===) so a NULL never matches "" or 0, and (for the
            // small number of columns where the two sides could disagree
            // on PHP type despite equal underlying data) a genuine type
            // difference still fails rather than being coerced away by ==.
            if ($originalValue === $replacementValue) {
                continue;
            }

            Assert::fail(sprintf(
                "ormharness: %s, column \"%s\" differs:\n  original:    %s\n  replacement: %s",
                $rowLabel,
                $c,
                self::formatValue($originalValue),
                self::formatValue($replacementValue)
            ));
        }
        Assert::assertTrue(true); // record an assertion per row so PHPUnit does not flag the test as risky
    }

    private static function formatValue(mixed $v): string
    {
        if ($v === null) {
            return 'NULL';
        }

        return sprintf('%s (%s)', var_export($v, true), get_debug_type($v));
    }
}
